raft/raft_log.cpp

```diff
@@ -0,0 +1,165 @@
+// Copyright (C) 2018 Bluzelle
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License, version 3,
+// as published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+#include "raft_log.hpp"
+
+#include <fstream>
+#include <boost/filesystem/operations.hpp>
+#include <iostream>
+
+
+namespace bzn
+{
+    raft_log::raft_log(const std::string& log_path) : entries_log_path(log_path)
+    {
+        if(boost::filesystem::exists(this->entries_log_path))
+        {
+            std::ifstream is(this->entries_log_path, std::ios::in | std::ios::binary);
+            bzn::log_entry log_entry;
+            while (is >> log_entry)
+            {
+                this->log_entries.emplace_back(log_entry);
+            }
+            is.close();
+        }
+
+        if (this->log_entries.empty())
+        {
+            throw std::runtime_error(MSG_ERROR_EMPTY_LOG_ENTRY_FILE);
+        }
+    }
+
+
+    const bzn::log_entry&
+    raft_log::entry_at(size_t i) const
+    {
+        return this->log_entries.at(i);
+    }
+
+
+    const bzn::log_entry&
+    raft_log::last_quorum_entry() const
+    {
+        // TODO: Speed this up by not doing a search, when a quorum entry is added, simply store the index. Perhaps only do the search if the index is wrong.
+        auto result = std::find_if(
+                this->log_entries.crbegin(),
+                this->log_entries.crend(),
+                [](const auto& entry)
+                {
+                    return (entry.entry_type == bzn::log_entry_type::single_quorum) || (entry.entry_type == bzn::log_entry_type::joint_quorum);
+                });
+        if(result == this->log_entries.crend())
+        {
+            throw std::runtime_error(MSG_NO_PEERS_IN_LOG);
+        }
+        return *result;
+    }
+
+
+    void
+    raft_log::append_log_disk(const bzn::log_entry& log_entry)
+    {
+        if (!this->log_entry_out_stream.is_open())
+        {
+            boost::filesystem::path path{this->entries_log_path};
+            if(!boost::filesystem::exists(path.parent_path()))
+            {
+                boost::filesystem::create_directories(path.parent_path());
```

Comment:
should we not check to see if the path is created?

Reply:
Done, I chose to use the non-throwing version of create_directories in this way:

boost::system::error_code ec;
if(!boost::filesystem::create_directories(path.parent_path(), ec))
{
    LOG(error) << "Unable to create path " << path.parent_path() << " with error code " << ec <<".";
    throw std::runtime_error(MSG_EXITING_DUE_TO_LOG_PATH_CREATION_FAILURE);
}

so that I didn't need to catch and re throw an exception